File: solutions/206.reverse-linked-list/src/recursion.rs
/*
 * @lc app=leetcode.cn id=206 lang=rust
 *
 * https://leetcode.cn/problems/reverse-linked-list/
 */

pub struct Solution;

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// @lc code=start
impl Solution {
    fn reverse_list_recursive(
        curr: Option<Box<ListNode>>,
        prev: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        match curr {
            Some(mut curr) => {
                let next = curr.next;
                curr.next = prev;
                Self::reverse_list_recursive(next, Some(curr))
            }
            None => prev,
        }
    }

    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        Self::reverse_list_recursive(head, None)
    }
}
// @lc code=end

File: solutions/206.reverse-linked-list/src/recursion.ts
/*
 * @lc app=leetcode.cn id=206 lang=typescript
 *
 * https://leetcode.cn/problems/reverse-linked-list/
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

// @lc code=start
function reverseListRecursive(
    curr: ListNode | null,
    prev: ListNode | null,
): ListNode | null {
    if (!curr) {
        return prev;
    }
    const next = curr.next;
    curr.next = prev;
    return reverseListRecursive(next, curr);
}

function reverseList(head: ListNode | null): ListNode | null {
    return reverseListRecursive(head, null);
}
// @lc code=end

export { reverseList };
